### PR TITLE
lua: update to 5.4.0

### PR DESCRIPTION
**Pull request by pekdemira**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.6
Xcode 11.6

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by mojca** [comment]
Updating lua may cause breaking of all dependent ports. We should create a port for the old version and check all dependent ports. If nothing else they need a revbump.

File: lang/lua/Portfile [hunk 1]
```diff
@@ -15,23 +15,36 @@ long_description    Lua is a powerful, light-weight programming language designe
...
+livecheck.type      regex
+livecheck.url       ${master_sites}
+livecheck.regex     {lua-(\d+(?:\.\d+)*)}
 
 # TODO: also add lua50 and lua51 here
+subport             lua53 {
+    version         5.3.5
+    revision        0
+    checksums       rmd160  f260c6f60139cb1464d87b5a7d9673e3f305f368 \
+                    sha256  0c2eed3f960446e1a3e4b9a1ca2f3ff893b6ce41942cf54d5dd59ab4b3b058ac
```
Comment (mojca) [inline lang/lua/Portfile:39]: Size is missing here.

File: lang/lua/Portfile [hunk 2]
```diff
@@ -15,23 +15,36 @@ long_description    Lua is a powerful, light-weight programming language designe
...
 
 patchfiles          patch-Makefile.diff \
                     patch-src-Makefile.diff \
                     patch-src-luaconf.h.diff
 
 # TODO: fixme (a different livecheck is needed for different subports)
-livecheck.type  regex
-livecheck.url   ${master_sites}
-livecheck.regex {lua-(\d+(?:\.\d+)*)}
+livecheck.type      regex
```
Comment (mojca) [inline lang/lua/Portfile:30]: Minor and not extremely important nitpicking. We try to make whitespace changes into a separate commit.

File: lang/lua/files/patch-src-Makefile.diff [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
 +++ src/Makefile
 @@ -7,7 +7,7 @@
- PLAT= none
+ PLAT= guess
  
  CC= gcc -std=gnu99
--CFLAGS= -O2 -Wall -Wextra -DLUA_COMPAT_5_2 $(SYSCFLAGS) $(MYCFLAGS)
-+CFLAGS= -O2 -Wall -Wextra -DLUA_COMPAT_5_1 -DLUA_COMPAT_5_2 $(SYSCFLAGS) $(MYCFLAGS)
+-CFLAGS= -O2 -Wall -Wextra -DLUA_COMPAT_5_3 $(SYSCFLAGS) $(MYCFLAGS)
++CFLAGS= -O2 -Wall -Wextra -DLUA_COMPAT_5_1 -DLUA_COMPAT_5_2 -DLUA_COMPAT_5_3 $(SYSCFLAGS) $(MYCFLAGS)
```
Comment (mojca) [inline lang/lua/files/patch-src-Makefile.diff:8]: I have my share of doubts that this helps in any way.

**Comment by pekdemira** [comment]
> Updating lua may cause breaking of all dependent ports. We should check all of them: if nothing else, they need a revbump.

You're answering a part of my question I tried to address on Trac @ #61007. Sounds like we want to create a new port for the old 5.3 version and update the dependencies accordingly.

**Comment by mojca** [comment]
We can continue the discussion here, there's not much benefit of spreading it in two different places.
There is no major difference between having a standalone port and a subport. They both work exactly the same way as far as MacPorts is concerned. But when ports share tons of the same options (same website, same base URL for downloading versions, same livecheck, almost he same description, ...) it's easier to maintain them if they are combined in the same port.

In the past we didn't have subports and creating a standalone port was the only option. Even after we introduced subports, many developers still weren't too familiar with the concept and didn't think of subports even when they could have used them.

I didn't actually check the logs of lua Portfile history, but I suspect that in the first few rounds whoever tried to make the old software work simply create a standalone lua port. When updating to 5.3 a subport was created (which kind of makes more sense), but the developer didn't take the time to "clean up" the old ports, and only left a short notice for any further developers running into this.

In my initial reply I didn't immediately noticed that you have already create `lua53` which was the correct thing to do. We just need to ensure that software depending on lua still builds. Each minor number increase (like 5.2 -> 5.3) usually introduces breaking changes that require fixing the software depending on lua. Enabling the compatibility mode mostly doesn't really help.

**Comment by pekdemira** [comment]
@mojca  - Agreed. Keeping it in one place is also beneficial for future reference. Got it closed on trac. Also, thank you for your note. I think now I understand how subports do work.

I have updated the branch after the review. As of this writing the build has passed. Once this branch merged, I may go ahead, clean up and add remaining versions to subports as well.

Only question I have is about the ability to test a port againsts its dependencies. if there is a framework to do it. Another one is that the port name `lua` currently pointing to different versions over time as updates are pushed. I wonder if that would have any possible side effects.

Thank you,

**Comment by kencu** [comment]
Do we push this, then revbump all the ports that use lua and fix wreckage? 

Or do we want to test them all first (a moderately cumbersome project). I presume the ones that don't build against lua 5.4 we move to use lua 5.3 as they are doing now.

**Comment by kencu** [comment]
What is needed to commit this?

**Comment by pekdemira** [comment]
I cleared the blockers at the time revising the code. It was ready to merge. Given the time elapsed, this needs to be revised again, and will need to be updated with the minor fixes in the latest release.

**Comment by kencu** [comment]
Another three weeks, no progress here. What is going on with this PR?

**Comment by pekdemira** [comment]
This PR is now out of date being six months old. Needs to be updated to the latest release.

Don't have the time right now. May visit back later this month unless somebody else taken care of it.

Closing this PR.
